modules/packages/DistributedBag.chpl

```diff
@@ -83,183 +94,191 @@
-
+module DistributedBag
+{
   public use Collection;
-  use BlockDist;
-  private use CTypes;
   private use IO;
 
+  use Random;
+  use List;
+  use Math;
+
   /*
-    Below are segment statuses, which is a way to make visible to outsiders the
-    current ongoing operation. In segments, we use test-and-test-and-set spinlocks
-    to allow polling for other conditions other than lock state. `sync` variables
-    do not offer a means of acquiring in a non-blocking way, so this is needed to
-    ensure better 'best-case' and 'average-case' phases.
+    The scenarios of the remove operation (see header).
   */
-  private param STATUS_UNLOCKED : uint = 0;
-  private param STATUS_ADD : uint = 1;
-  private param STATUS_REMOVE : uint = 2;
-  private param STATUS_LOOKUP : uint = 3;
-  private param STATUS_BALANCE : uint = 4;
+  private param REMOVE_BEST_CASE    = 1;
+  private param REMOVE_LOCAL_STEAL  = 2;
+  private param REMOVE_GLOBAL_STEAL = 3;
+  private param PERFORMANCE_PATCH   = 4;
 
   /*
-    Below are statuses specific to the work stealing algorithm. These allow the
-    shepherd tasks to know when its sub-helpers finish and the end status of their
-    work stealing attempt.
+    Outputs of the remove operation. They are used to indicate the final status
+    of the operation: success or fail.
   */
-  private param WS_INITIALIZED = -1;
-  private param WS_FINISHED_WITH_NO_WORK = 0;
-  private param WS_FINISHED_WITH_WORK = 1;
+  private param REMOVE_SUCCESS   = true;
+  private param REMOVE_FAIL      = false;
+  // NOTE: A third output could be used to optimize user code in scenarios
+  // where the remove operation fails, but there is a pending work stealing
+  // operation - meaning that the DistBag is not necessarily empty.
+  // For now, it is considered as a usual fail.
+  private param REMOVE_FAST_EXIT = REMOVE_FAIL;
 
   /*
-    The phases for operations. An operation is composed of multiple phases,
-    where they make a full pass searching for ideal conditions, then less-than-ideal
-    conditions; this is required to ensure maximized parallelism at all times, and
-    critical to good performance, especially when a node is oversubscribed.
+    The initial capacity of each segment. When a segment is full, we double its
+    capacity.
   */
-  private param ADD_BEST_CASE = 0;
-  private param ADD_AVERAGE_CASE = 1;
-  private param REMOVE_BEST_CASE = 2;
-  private param REMOVE_AVERAGE_CASE = 3;
-  private param REMOVE_WORST_CASE = 4;
-
+  config const distributedBagInitialSegmentCap: int = 1024;
   /*
-    The initial amount of elements in an unroll block. Each successive unroll block
-    is double the size of it's predecessor, allowing for better locality for when
-    there are larger numbers of elements. The better the locality, the better raw
-    performance and easier it is to redistribute work.
+    The maximum capacity of each segment. This is crucial to ensure memory usage
+    does not grow out of control.
```

Comment:
If the limit is reached, we cease to offer more, and return the number of elements successfully inserted. It means that some elements may be missed. This behaviour is inherited from the old data structure, itself inherited from the `Collection` module.

This behaviour is described in the doc of the `addBulk` proc:

>  Insert elements in bulk in segment ``taskId``. If the bag instance rejects
      an element (e.g., if :const:distributedBagMaxSegmentCap is reached), we
      cease to offer more. We return the number of elements successfully inserted.

and in that case, some warnings are generated at runtime:

> warning("maximum capacity reached: some elements may have been missed.");

Reply:
OK. IMO it is preferable to rely on the return value from `add` and `addBulk` instead of the runtime warning for this. See also my comment https://github.com/chapel-lang/chapel/pull/23204/files#r1412532304 .